### ISSUE TITLE
solidjs style computeds

### ISSUE THREAD

**Issue by btakita** [nanostores/nanostores#188]
Solidjs signals use inline dependencies, removing the redundancy of declaring the dependency atoms when the computed is defined. I think we can lazily add dependencies when the computed function is executed.

Perhaps it could look something like:

```ts
const base = atom(0)
const dependency = computed(()=>base() + 1)
console.info({ dependency: dependency() })
dependency.unbind()
```

**Comment by ai** [comment]
It looks interesting. Can somebody make an PR? We need to check the size of this feature.

**Comment by btakita** [comment]
I can work on a PR when I have some free time, which will be at least 2 weeks from now. At the very least, this can be a separate package, but it may be small enough to include in this project. To anyone reading this who wants to work on this, don't let this message dissuade you from creating a patch.

**Comment by btakita** [comment]
Hey @dtinth! I created [a PR](https://github.com/nanostores/nanostores/pull/203) which adds autosubscribe & async support. So your example would be supported as you wrote it using `computed` or since atoms are functions, can be called directly with autosubscribe in synchronous handlers.

```ts
const $shape = atom('circle')
const $radius = atom(4)
const $width = atom(2)
const $height = atom(3)

const $area = computed(() => {
  if ($shape() === 'circle') {
    return Math.PI * $radius() ** 2
  }
  if ($shape() === 'rectangle') {
    return $width() * $height()
  }
})
```

```ts
const $a = atom('a1')
const $b = atom('b1')
const $switcher = atom($a)
const $c = computed(() => {
  return $switcher()()
})
```

There is also support for async computeds. Re-entrancy is also supported, meaning if a callback is run before a prior callback completes, the prior callback will be marked as "stale". An example:

```ts
computed(async use => {
  let person = await fetch(`https://api.com/people/${personId()}`).then(response => response.json())
  if (use.stale()) return // stop run if the cb is stale
  let messages = await fetch(`https://api.com/people/${personId(use)}/messages`).then(response => response.json())
  return { person, messages } // This will not be saved if the cb is stale
})
```

**Comment by btakita** [comment]
> I am impressed with the implementation of async. I think it would be useful, especially that you handled the "stale" case.

Thank you! There are some [additional use cases](https://github.com/btakita/nanostores/blob/issues/188/computed/index.d.ts#L17-L129) which includes **Event binding & unbinding** and **Helper functions & Private stores**.

The helper functions on `Autosubscribe` (`use`) are:

- `on(fn: () => any): Autosubscribe<Value>`: callback when the computed is first listened to. Useful for binding to events within the computed cb
- `off(fn: () => any): Autosubscribe<Value>`: callback when all listeners are unbound. Useful for unbinding to events within the computed cb
- `set(intermediateValue: Value): Autosubscribe<Value>`: sets a intermediate value of the computed. Useful for loading states & other intermediate states which can be rolled back to the last saved value using `.undo()`
- `save(newValue: Value): Autosubscribe<Value>`: sets the computed value & `undoValue`
- `undo(): Autosubscribe<Value>`: resets the computed back to the last saved value
- `stale(): boolean`: returns true when the computed's cb is called after the current run

> I’m not a fan of using () to read the store value though, because then there are 2 ways to read the value. I am afraid that it could lead to inconsistent API usage. Regarding the TypeScript inference issue, I think it’s possible to do this: cb: (get: <T>(atom: ReadonlyStore<T>)=>T) => Value (this is what I do in my lib).

First of all, thank you for the type inference tip. I can confirm that it works & now the `use` function correctly infers the return type.

I can see your point re: keeping the interface small. Another benefit to not having an atom as a function is that the `autosubscriberStack` & `autosubscriber` are not necessary. This would reduce the size of the implementation & simplify the api. I agree that it is more appropriate to have `use` be responsible for autosubscription.

I personally like having `atom` be a function because it reduces the noise of overusing `use`, which is particularly helpful when multiple stores are autosubscribed in close proximity. I also have multiple libraries with functions that extend `atom` & `computed`. In this case, I can extend `atom` & `computed` via a `Proxy` to give an atom a function interface.

I prefer having the api sugar to improve readability of my own projects, but I also would rather not have to document & maintain multiple ways of autosubscribing in the nanostores project...I also have the sense that others, including @TrySound would also prefer the simplicity of only supporting `use` and the library size would also be less impacted.

So I appreciate your comments & am adjusting the PR accordingly. If anyone wants to use the `atom` function interface, I can extract a separate library which does this.

**Comment by btakita** [comment]
I removed the `ReadableAtom` function api in the [btakita/nanostores#issues/188](https://github.com/btakita/nanostores/tree/issues/188) branch used by the [PR 203](https://github.com/nanostores/nanostores/pull/203).

I also have the [btakita/nanostores#issues/188-atom-fn](https://github.com/btakita/nanostores/tree/issues/188-atom-fn) available which keeps the `ReadableAtom` function api.

If we still prefer to have functionless atoms in nanostores, I can extract a new package for the function atoms. Either way, it would be great to have the PR merged.

**Comment by ai** [comment]
Sorry, I still have no time to review that PR. I need to prepare a talk and PR is very big to review it fast.

I will try to do it in July.

**Comment by mindplay-dk** [comment]
I thought auto subscribe was a non-feature by design?

I personally want to see the list of dependencies as the first argument whenever a computed is declared - I find this approach to be more explicit and declarative.

I don't have to read through the entire implementation of the function body to figure out which dependencies the function has - this is a good thing.

It's the first principle of dependency injection: "don't look for things" - your function has no business reaching out directly for it's dependencies. "ask, don't tell" and so forth. 

Also, on terms of simplicity, things are much easier to understand without the global state and side effects required to implement this run-time instrumentation of dependencies - the library is self explanatory, as it is.

> Solidjs signals use inline dependencies, removing the redundancy of declaring the dependency atoms when the computed is defined

Solid is just one of the frameworks supported by this library. I don't think this is a great argument for changing the fundamental design of this library.

In fact, with this change, I would question why you'd even need or want this library in Solid? It would work essentially just like the state management that's already built into Solid. Just use that. 

I took an interest in this library *because* it's based on good design principles, and because it's so simple and easy to explain. 

There are already several other libraries with Solid style signals, which have automatic dependency tracking, such as [Legend](https://legendapp.com/open-source/state/v3/) or [Preact Signals](https://preactjs.com/guide/v10/signals/), not to mention the official [Solid Signals](https://github.com/solidjs/signals) by Ryan himself, which is currently in development.

There is only one library like `nanostores`, as far as I'm aware - if this becomes just yet another implementation of signals, this breed of state management goes extinct.

I hope you don't. 😿

Diversity is good for the ecosystem - everything doesn't have to converge to the same thing every time something gets popular. 😌

**Comment by dtinth** [comment]
@mindplay-dk 

> “I find this approach to be more explicit and declarative”

I agree that it is more explicit. However, the current “static dependency list” design makes it impossible to declare a computed atom whose dependencies that should be tracked can dynamically change. I worked around this by publishing [nanostores-computed-dynamic](https://github.com/dtinth/nanostores-computed-dynamic) but it’s not maintained and IMO I’d rather have such functionality built in to nanostores.

> “I would question why you'd even need or want this library in Solid?”

I agree that you wouldn’t need this library if you only build Solid apps — you can just use Solid’s built-in signals. But sometimes I work with multiple apps and they were written using different frameworks. Nano Stores is one of the few state management libraries that have great support for all popular frameworks. So nowadays I tend to write the core logic in Nano Stores so it can be agnostic of UI framework. I think lack of dynamic dependency tracking is adding some friction to creating performant Nano Stores-based app.

**Comment by mindplay-dk** [comment]
> I agree that it is more explicit. However, the current “static dependency list” design makes it impossible to declare a computed atom whose dependencies that should be tracked can dynamically change ([I elaborated about the need for “dynamic dependency tracking” back in 2018](https://stackoverflow.com/a/51973044)). I worked around this by publishing [nanostores-computed-dynamic](https://github.com/dtinth/nanostores-computed-dynamic) but it’s not maintained and IMO I’d rather have such functionality built in to nanostores.

I agree, but runtime tracking probably isn't the right way - we need to think of a functional, declarative way to achieve this, like how everything else is achieved with the existing API.

Based on your example in `nanostores-computed-dynamic`, how about something like:

```js
const $shape = atom('circle')
const $radius = atom(4)
const $width = atom(2)
const $height = atom(3)

const $area = branch($shape, {
  circle: () => computed([$radius], (radius) => Math.PI * use($radius) ** 2),
  rectangle: () => computed([$width, $height], (width, height) => width * height)
});
```

Each time the branch control `$shape` updates, the branch function would tear down the previous `computed` and set up a new one. I think that's exactly how tracked signals work, except that tracked signals have to tear down an redetermine the dependencies after every update. (performance wise, this approach might actually be slightly better since it only needs to establish new dependencies when the branch control actually changes.)

In other words, the branch control explicitly controls branch switching - only one branch is active at a time, with its own specific dependencies. Clean teardown/setup whenever the branches switch, and so on.

Since a `branch` is just a specific kind of `computed`, this approach could probably work with conditional branching as well:

```js
const $view = branch($page, {
  dashboard: () => branch($dashboardTab, {
    overview: () => computed([$metrics], metrics => ...),
    details: () => computed([$details], details => ...)
  }),
  settings: () => computed([$settings], settings => ...)
});
```

For boolean or other conditions, where some logic is required to pick the branch, just use a `computed` branch control:

```js
const $display = branch(
  computed([$darkMode, $size], (darkMode, size) => `${darkMode}-${size}`),
  {
    'true-small': () => computed(...),
    'true-large': () => computed(...),
    'false-small': () => computed(...),
    'false-large': () => computed(...)
  }
);
```

I think this more in tune with the existing framework? It's an addition to the existing API, rather than an entirely new approach - existing projects can adopt this incrementally, no need to refactor to an entirely new API style.

Dependencies remain explicit in every branch, no need for runtime tracking or complexity of determining dependencies on each update, and the branching itself stays declarative - the shape of the dependency graph is clear from reading the declarations, without having to read through the code to discover the dependencies.

**Comment by mindplay-dk** [comment]
Full disclosure: no clear idea what I'm doing here (Claude helped) but here's an attempt at a `branch` function as described in my comment:

https://github.com/nanostores/nanostores/compare/main...mindplay-dk:nanostores:branching

It's passing the test but for sure needs lots of additional testing. I'm not sure how to verify in the test that branches are correctly unsubscribed... I thought I was doing that indirectly by pushing to `log` when the two branch `computed` functions run - I thought, if the previous branch was still subscribed, you would see a log entry from the function that shouldn't be running but... removing the call to `branchUnsubscribe()` for some reason still passes the test?! which tells me that... I don't really know what I'm doing!? haha. 😅

**Comment by mindplay-dk** [comment]
It turns out, `branch` is just a special case of a simpler, more general, more flexible function - perhaps suited for more complex use case, or just for those who prefer a more procedural style:

```js
const $shape = atom('circle')
const $radius = atom(4)
const $width = atom(2)
const $height = atom(3)

const $area = fork($shape, (shape) => {
  if (shape === "circle") {
    return computed([$radius], (radius) => Math.PI * use($radius) ** 2)
  }
  if (shape === "rectangle") {
    return computed([$width, $height], (width, height) => width * height)
  }
  throw new Error(`unsupported shape: ${shape}`);
});
```

I added some more tests and fixed some bugs, but I still can't figure out why the previous `computed` seems to keep computing after I've unsubscribed.

I can't come up with a test that points to me doing something dumb, so I added `console.log` statements to the function, which you can see in the console when running the test... and it looks correct to me?

I am definitely unsubscribing, so it's not my code polling the inactive branch for updates, I don't think... so it must be the framework? I suspect maybe this has something to do with the 1000 msec delay before stores get cleaned up? I have no idea.

Anyways, I spent a day on this, and frankly I don't think I'm smart enough to solve this haha 😅

But what do you think, does it look like a viable alternative to "going full signals"?

(and sorry for hijacking this thread, I didn't intend for it to go this far...)

**Comment by mindplay-dk** [comment]
again, sorry for hijacking this thread, I've opened a separate issue #345 to discuss my idea. 🙂

**Comment by mindplay-dk** [comment]
> I think lack of dynamic dependency tracking is adding some friction to creating performant Nano Stores-based app.

In terms of performance, what I'm proposing in #345 would most likely perform better than dynamic dependency tracking.

From what I know, most implementations of signals all work on the principle of tearing down all subscriptions every time a computed is updated, instrumenting the dependencies while the compute function runs, and then resubscribing. At least, that's how I've heard Ryan Carniato explain it on his streams. (I'm not sure how the proposed implementation here works?)

There is also the issue of needing to call the computed function to initialize it's dependencies - I'm not sure how lazy `nanostores` is at the moment, but (as far as I can figure) you can't have have lazy computed signals unless the dependencies are provided statically.

If performance is a concern, run-time instrumentation of dependencies probably doesn't help.

**Comment by kubk** [comment]
Hi @ai, I'd love to revive the idea of dynamic dependency tracking. The previous PR was large but we could make a simpler version:
  - a separate function and package under the `@nanostores` scope. Smth like `computedAuto`
  - only supports sync calculations

  ```ts
  import { computedAuto } from '@nanostores/computed-auto'
  import { computed } from 'nanostores'

 // Before
  const $pullRequestStatusBadge = computed(
    [
      $isDraft,
      $pullRequestStatus,
      $checks,
      $review,
      $deployment,
    ],
    (isDraft, status, checks, review, deployment) => {
      if (isDraft) return 'Draft'
      if (status === 'closed') return 'Closed'
      if (checks === 'failed') return 'Checks failing'
      if (review === 'changes') return 'Changes requested'
      return deployment === 'deployed' ? 'Published' : 'Ready to merge'
    }
  )

  // After
  const $pullRequestStatusBadge = computedAuto(get => {
    if (get($isDraft)) return 'Draft'
    if (get($pullRequestStatus) === 'closed') return 'Closed'
    if (get($checks) === 'failed') return 'Checks failing'
    if (get($review) === 'changes') return 'Changes requested'
    return get($deployment) === 'deployed' ? 'Published' : 'Ready to merge'
  })
```

The code above is formatted by Prettier

Reasoning:
- Dynamic derivations can be more performant thanks to dynamic branching. Less recalculations and rerenders
- they're easier to read for a human, faster to write for an agent
- People coming from Solid, Svelte, MobX, or Vue would expect derivations to track dependencies automatically, not something a developer needs to specify manually
- A separate package will keep the core lean

If the direction makes sense I'd love to proceed with the contribution 👍

**Comment by ai** [comment]
@kubk very interesting! Any idea how we can block direct `$store.get()` inside `computedAuto()`?

ESLint rule nowadays is not so useful since not everyone reads docs to know that they need to install it.

For instance, we can add development-only check in `get()` for some marker set by `computedAuto` to throw error in development envrionment.

**Comment by ai** [comment]
@kubk btw, how `computedAuto` will look like?

**Comment by dtinth** [comment]
[3 years ago](https://github.com/nanostores/nanostores/issues/188#issuecomment-1611611341), I published a simple version of this proposal at [nanostores-computed-dynamic](https://github.com/dtinth/nanostores-computed-dynamic). I have been using this library since, but haven't got time to catch it up with latest nanostores versions. Happy to see this getting revived.

- My version of `computedAuto()`, [`computedDynamic()`, is just 80 lines of code](https://github.com/dtinth/nanostores-computed-dynamic/blob/main/src/index.ts). In my opinion, I think even if `computedAuto` goes into core, it wouldn't result in library bloat that much. Furthermore, it is completely tree-shakable.
- [Here's the test suite.](https://github.com/dtinth/nanostores-computed-dynamic/blob/main/src/test.ts) It covers quite a number of edge cases, in case it helps with the official implementation of `computedAuto()`.

> Any idea how we can block direct `$store.get()` inside `computedAuto()`?

I find myself using `$store.get()` directly inside `computedDynamic()` sometimes. It gives flexibility for the function to "read the current value of the store without subscribing."

- `get($store)` → read and subscribe
- `$store.get()` → read without subscribing

**Comment by ai** [comment]
> is just 80 lines of code

I like it =^_^=

Now we need a migration plan.

1. Keep `computed` and `computedAuto` together for a long time
2. Or add `computedAuto` in next minor and remove `computed` in next major

Any arguments for the way?

**Comment by kubk** [comment]
Hi @dtinth awesome work!

Since you already have a working version of this approach - maybe it makes sense to continue from there. If you're interested in continuing with the PR, feel free to ping me there. I am maintaining [MobX](https://github.com/mobxjs/mobx), its `computed` handles a lot of complex stuff already

**Comment by kubk** [comment]
@ai 

> Or add `computedAuto` in the next minor and remove `computed` in the next major

Seems like a good option! My only concern is plain JS users who upgrade to a new major without knowing that the behavior of `computed` changed. If the API changes from `computed(args, cb)` to `computed(cb)` their old calls will fail at runtime with cryptic error. So maybe this dev-only protection makes sense in the new major:

```ts
if (process.env.NODE_ENV !== 'production' && typeof args[0] !== 'function') {
  throw new TypeError(
    'computed() now expects a callback. Use computed(get => ...)'
  )
}

**Comment by psd-coder** [comment]
Hey guys! I really like the direction we're moving in here. Manually defining dependencies is not only tedious and error-prone, but also suboptimal. It might be a good feature for Nanostores v2.

But I believe we need to change not only `computed`, but also `effect`. Also, there are users who prefer to list dependencies manually. We might keep the old methods for them: it costs us nothing, except for a slightly more bloated API, maybe.

In my head, the ideal picture looks like this:
1. Introduce new `computedAuto` and `effectAuto` methods in a minor version.
2. Battle-test them a bit in our projects.
3. Release Nanostores v2, where `computedAuto` and `effectAuto` become the default `computed` and `effect`, with the assertion suggested by @kubk. The old-behavior methods might be kept as `computedManual` and `effectManual`.

**Comment by kubk** [comment]
@dtinth are you planning to open a PR or should I take it? i'm happy either way, but it makes sense to avoid duplicated work

**Comment by dtinth** [comment]
@kubk Please feel free to go ahead. I would love to but I am too busy with other priorities right now. Also, thanks for maintaining mobx! I am using it in some of my projects too.

**Comment by ai** [comment]
I do not like the `computedAuto` naming in long term. What do you think if the auto-mode will be just an extent of old manual mode (with abilityt to mix)?

```js
// Adding extra argument
$one = computed([$two, $three], (two, three, get) => {
  return two + three + get($four)
})
// On full auto mode
$one = computed([], (get) => {
  return get($two) + get($three) + get($four)
})
// Shortcut
$one = computed((get) => {
  return get($two) + get($three) + get($four)
})
```

**Comment by psd-coder** [comment]
I'd avoid supporting second option. I read it as computed with no dependencies (set by mistake, or somewhat)
```
// On full auto mode
$one = computed([], (get) => {
  return get($two) + get($three) + get($four)
})
```
Only first or last works better, IMO. I prototyped it a bit.
